### PR TITLE
Fix #971: Implement auditing for Next Step identity

### LINKED ISSUE #971: Implement auditing for Next Step identity
Once auditing functionality is available in lime-java-core, implement it in Next Step.

See: https://github.com/wultra/lime-java-core/issues/34

### PR DESCRIPTION
**Pull request by romanstrobl**

### CODE REVIEW

File: powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/exception/DefaultExceptionResolver.java [hunk 1]
```diff
@@ -56,6 +72,7 @@
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public @ResponseBody ErrorResponse handleDefaultException(Throwable t) {
         logger.error("Error occurred in Next Step server", t);
+        audit.error("Error occurred in Next Step server", AUDIT_DETAIL_UNEXPECTED_ERROR, t);
```
Comment (petrdvorak) [inline powerauth-nextstep/src/main/java/io/getlime/security/powerauth/app/nextstep/exception/DefaultExceptionResolver.java:75]: I am not sure if we want to audit all exceptions - there are already stored in logger appender and do not represent the business logic situation as much as application errors. The purpose of audit log is to store business audit (what the user was doing in the system), not technical logging.

File: powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/errorhandling/DefaultExceptionResolver.java [hunk 2]
```diff
@@ -52,6 +68,7 @@
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public @ResponseBody ErrorResponse handleDefaultException(Throwable t) {
         logger.error("Error occurred in TPP engine server", t);
```
Comment (petrdvorak) [inline powerauth-tpp-engine/src/main/java/io/getlime/security/powerauth/app/tppengine/errorhandling/DefaultExceptionResolver.java:70]: ... same as above.

File: powerauth-webflow-authentication-mtoken/src/main/java/io/getlime/security/powerauth/lib/webflow/authentication/mtoken/errorhandling/MobileApiExceptionResolver.java [hunk 3]
```diff
@@ -44,13 +47,28 @@
...
+     */
+    @Autowired
+    public MobileApiExceptionResolver(Audit audit) {
+        this.audit = audit;
+    }
+
+
     private ErrorResponse error(String code, Throwable t) {
         logger.warn("Error occurred in Mobile Token API component", t);
+        audit.warn("Error occurred in Mobile Token API component", AUDIT_DETAIL_BAD_REQUEST, t);
```
Comment (petrdvorak) [inline powerauth-webflow-authentication-mtoken/src/main/java/io/getlime/security/powerauth/lib/webflow/authentication/mtoken/errorhandling/MobileApiExceptionResolver.java:65]: dtto

File: powerauth-webflow-authentication/src/main/java/io/getlime/security/powerauth/lib/webflow/authentication/exception/AuthenticationExceptionResolver.java [hunk 4]
```diff
@@ -55,6 +70,7 @@
     @ResponseStatus(HttpStatus.BAD_REQUEST)
     public @ResponseBody ErrorResponse handleMethodNotValidException(AuthStepException ex) {
         logger.warn("Error occurred in Web Flow server: {}", ex.getMessage());
+        audit.warn("Error occurred in Web Flow server: {}", AUDIT_DETAIL_BAD_REQUEST, ex);
```
Comment (petrdvorak) [inline powerauth-webflow-authentication/src/main/java/io/getlime/security/powerauth/lib/webflow/authentication/exception/AuthenticationExceptionResolver.java:73]: dtto

File: powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/exception/DefaultExceptionResolver.java [hunk 5]
```diff
@@ -61,6 +70,7 @@ public DefaultExceptionResolver(AuthenticationManagementService authenticationMa
     @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
     public @ResponseBody ErrorResponse handleDefaultException(Throwable t) {
         logger.error("Error occurred in Web Flow server", t);
+        audit.error("Error occurred in Web Flow server", AUDIT_DETAIL_UNEXPECTED_ERROR, t);
```
Comment (petrdvorak) [inline powerauth-webflow/src/main/java/io/getlime/security/powerauth/app/webflow/exception/DefaultExceptionResolver.java:73]: dtto

**Review by petrdvorak — CHANGES_REQUESTED** [review]
I am not sure if we should replicate logging into audit log for everything that we log. Specifically, any technical issues of our service should probably not go into the audit log.

The purpose of the audit log is to store **business relevant data**, basically the user activity in our system. The user of the data in this table is a non-technical worker responsible for ad-hoc task investigation.

**Comment by romanstrobl** [comment]
In general I do agree that technical messages should not appear in audit records. However my issue with not saving the error records is that the audit trail would not provide enough information in case there is an unexpected error - the audit trail stops suddenly. But I do not feel strongly about this, we can either remove the error audits pr switch them to debug (prefered by me). As experienced many times when investigating client issues more information is usually better as long as we do not waste DB space.

**Comment by romanstrobl** [comment]
As discussed offline, we will keep the error audit records for now due to their importance during investigation of what happened.

**Pull request merged**